### ISSUE TITLE
IDE0039 potential false-positive

### ISSUE THREAD

**Issue by yaakov-h** [dotnet/roslyn#23603]
**Version Used**: 
Visual Studio 2017 Enterprise
15.4
15.5

**Steps to Reproduce**:

Write out the following code in the editor:

```cs
using System;

namespace ScratchForRoslyn
{
	public class Class1
	{
		static void DoSomething(int i)
		{
			Func<object> func = () => SomeFunction();

			if (i < 0)
			{
				Foo(func);
			}

			if (i < 5)
			{
				Bar(func);
			}

			if (i < 10)
			{
				Baz(func);
			}
		}

		static void Foo(Func<object> o) => throw null;
		static void Bar(Func<object> o) => throw null;
		static void Baz(Func<object> o) => throw null;

		static object SomeFunction() => throw null;
	}
}
```

**Expected Behavior**:

No warnings, errors, or suggestions.

**Actual Behavior**:

IDE0039 suggests to change line 9 to a local function, like so:

```cs
object func() => SomeFunction();
```

This subtle code change results in three `Func<>` objects being created at runtime, rather than just one.

**Comment by CyrusNajmabadi** [comment]
Right.  It's a suggestion :)  Something you can choose to do if you think its an improvement.  It's not meant to exactly preserve exact runtime semantics.

**Comment by yaakov-h** [comment]
Ah, ok. I'm used to these suggestions being purely syntactical changes (e.g. out var, expression-bodied methods), so I was quite surprised when it was pointed out to me that this change actually has a (mild) performance impact.
